### ISSUE TITLE
clarify pull from docker registries

### ISSUE THREAD

**Issue by tbugfinder** [hpc/charliecloud#822]
Hello,

I am looking into charliecloud in order to support it within Nextflow - https://github.com/nextflow-io/nextflow/issues/1707.
I started off with the command reference however it looks like charliecloud has a direct dependency to an external pull command (docker pull, skopeo, ff).
As the CLI reference doesn't include ch-pull I'm wondering if this is a target you are looking for, or if we should go for using e.g. "modern" skopeo approach.

Thanks & Best

**Comment by reidpr** [comment]
Hi, thanks for getting in touch! That's great news.

TL;DR: Charliecloud does have an internal puller now.

I think the Charliecloud concept you are after is the *builder*. Charliecloud has wrappers for three builders — Docker, Buildah, and our internal `ch-grow` — but is able to run any container presented as a directory. This is, I believe, best documented on the `ch-build` [man page](https://hpc.github.io/charliecloud/command-usage.html#ch-build), which isn't the best place but it's what we have for now. FWIW, here is our [road map](https://github.com/hpc/charliecloud/projects/1) for the internal, fully unprivileged builder `ch-grow`.

So, to pull, you need to use a pull that corresponds to your builder. For Docker that's `docker pull` of course, but we *do* have a command `ch-tug` that will pull the image to `ch-grow` storage (note, I suspect `ch-tug` will be renamed `ch-grow --pull` soon). This depends on the Requests and lark-parser Python modules, but that's it. In the past we did depend on skopeo and umoci, but that was dropped a few releases ago.

Does that help?

Let me know what else we can do to assist your work.

**Comment by tbugfinder** [comment]
Thank you, @reidpr .
To start off, I don't want to build images from Dockerfiles but pull images from e.g. Dockerhub, AWS ECR or local registries (e.g. Nexus). Therefore `ch-tug` sounds like the command to start off with.

**Comment by reidpr** [comment]
Great! It sounds like your question is answered, so I'll close this issue. However, please do re-open it if I've misunderstood or answered incompletely.

**Comment by reidpr** [comment]
Charliecloud doesn't use `docker://`. Details on the image reference format are in our [FAQ](https://hpc.github.io/charliecloud/faq.html#how-do-i-specify-an-image-reference) (aside, I believe this to be the most complete documentation for that available anywhere).

`godlovedc/lolcow` would be a correct image reference and it would look on Docker Hub. It if were on private registry `foo.com:8080`, you would write `foo.com:8080/godlovedc/lolcow` I believe.

**Comment by tbugfinder** [comment]
Thank you! I missed that piece of the documentation. It would be really helpful if search wouldn't end up in 404 :-) 

Anyway, it looks like silent authentication (e.g. using ENV variables) isn't yet supported.
https://github.com/hpc/charliecloud/blob/master/lib/charliecloud.py#L701-L702

**Comment by reidpr** [comment]
> Thank you! I missed that piece of the documentation. It would be really helpful if search wouldn't end up in 404 :-)

Indeed. We'll increase the priority. Thanks for pointing it out (again)!

> Anyway, it looks like silent authentication (e.g. using ENV variables) isn't yet supported.
> https://github.com/hpc/charliecloud/blob/master/lib/charliecloud.py#L701-L702

Also true. Can you file a new feature request issue describing how you'd want it to work? It seems that "silent authentication" isn't Docker jargon? Environment variables would be pretty straightforward. Working with a credentials store would be harder.

**Comment by reidpr** [comment]
@tbugfinder Looks like this isn't quite resolved — can you clarify "silent authentication"?

**Comment by tbugfinder** [comment]
sorry, I missed that.
Basically I thought about using pre-defined environment variables and just apply those during the registry authentication process.
```
CH_DOCKER_USERNAME=myregistryusername
CH_DOCKER_PASSWORD=password
```
The name DOCKER might be misleading as it could be different hosting or registry type, though.
```
CH_OCI_USERNAME=myregistryusername
CH_OCI_PASSWORD=password
```

**Comment by reidpr** [comment]
Great! I pulled that into a separate feature request issue. Given that, do we need to keep this question open still?
